cmd/subctl/join.go

```diff
@@ -0,0 +1,238 @@
+	cmd.Flags().UintVar(&joinFlags.GlobalnetClusterSize, "globalnet-cluster-size", 0,
+		"cluster size for GlobalCIDR allocated to this cluster (amount of global IPs)")
+	cmd.Flags().StringVar(&joinFlags.GlobalnetCIDR, "globalnet-cidr", "",
+		"GlobalCIDR to be allocated to the cluster")
+	cmd.Flags().StringSliceVar(&joinFlags.CustomDomains, "custom-domains", nil,
+		"list of domains to use for multicluster service discovery")
+	cmd.Flags().StringSliceVar(&joinFlags.ImageOverrideArr, "image-override", nil,
+		"override component image")
+	cmd.Flags().BoolVar(&joinFlags.HealthCheckEnable, "health-check", true,
+		"enable Gateway health check")
+	cmd.Flags().Uint64Var(&joinFlags.HealthCheckInterval, "health-check-interval", 1,
+		"interval in seconds between health check packets")
+	cmd.Flags().Uint64Var(&joinFlags.HealthCheckMaxPacketLossCount, "health-check-max-packet-loss-count", 5,
+		"maximum number of packets lost before the connection is marked as down")
+	cmd.Flags().BoolVar(&joinFlags.GlobalnetEnabled, "globalnet", true,
+		"enable/disable Globalnet for this cluster")
+	cmd.Flags().StringVar(&joinFlags.CorednsCustomConfigMap, "coredns-custom-configmap", "",
+		"Name of the custom CoreDNS configmap to configure forwarding to lighthouse. It should be in "+
+			"<namespace>/<name> format where <namespace> is optional and defaults to kube-system")
+	cmd.Flags().BoolVar(&joinFlags.IgnoreRequirements, "ignore-requirements", false, "ignore requirement failures (unsupported)")
+}
+
+func askForGatewayNode(allWorkerNodeNames []string) (struct{ Node string }, error) {
+	qs := []*survey.Question{
+		{
+			Name: "node",
+			Prompt: &survey.Select{
+				Message: "Which node should be used as the gateway?",
+				Options: allWorkerNodeNames,
+			},
+		},
+	}
+
+	answers := struct {
+		Node string
+	}{}
+
+	err := survey.Ask(qs, &answers)
+	if err != nil {
+		return struct{ Node string }{}, err // nolint:wrapcheck // No need to wrap here
+	}
+
+	return answers, nil
+}
+
+func checkArgumentPassed(args []string) error {
+	if len(args) == 0 {
+		return errors.New("broker-info.subm file generated by 'subctl deploy-broker' not passed")
+	}
+
+	return nil
+}
+
+func askForClusterID() (string, error) {
+	// Missing information
+	qs := []*survey.Question{}
+
+	qs = append(qs, &survey.Question{
+		Name:   "clusterID",
+		Prompt: &survey.Input{Message: "What is your cluster ID?"},
+		Validate: func(val interface{}) error {
+			str, ok := val.(string)
+			if !ok {
+				return nil
+			}
+
+			_, err := join.IsValidClusterID(str)
+			fmt.Printf("error is %s", err)
+			return err
+		},
+	})
+
+	answers := struct {
+		ClusterID  string
+	}{}
+
+	err := survey.Ask(qs, &answers)
+	// Most likely a programming error
+	if err != nil {
+		return "", err
+	}
+
+	return answers.ClusterID, nil
+}
+
+func askForCIDR(name string) (string, error) {
+	autodetect := false
+	prompt := &survey.Confirm{
+		Message: fmt.Sprintf("Do you want to autodetect %s CIDR?", name)}
+
```

Comment:
This is a change in behavior from the current `join` command where it tried to discover the network details and prompt the user if needed. Since the intention of this PR is to restructure the code, I think we should keep the same behavior. So that means the network discovery should be initiated here and not in the library component.

Reply:
I think the library should autodetect the networks if no information is passed. So this is what is implemented here. For third party user, it can't be prompted to enter network details if autodetection fails. And hence these changes. The design is
"if network details are not provided via subctl flags, prompt the user. if the details are provided but doesnot match the autodetected, use the provided details." Library always expects the some or empty (strings) details to be provided.

Reply:
Proposed changes with https://github.com/submariner-io/submariner-operator/pull/1726/commits/ce48e99552685255569ecf3c7a7e84ea4b2a5e63.